### DISCUSSION TITLE
Support for batching transactions in SQLite writer

### DISCUSSION THREAD

**Discussion by AdrianVollmer** [zeek/zeek#1901] [Ideas]
I'm trying to parse a 1GB pcap and write it into sqlite databases. It takes something around half an hour, whereas the ASCII writer takes about 10 seconds for the same file. I'm planning on parsing lots of those files, so performance is important.

It's not clear to me what the reason could be. CPU usage of the zeek process is around 10%. Memory usage is nowhere near the maximum.

Judging from the [source](https://github.com/zeek/zeek/blob/ff98515f2aa94985f16dbc08ceb1d5ba76b92ddb/src/logging/writers/sqlite/SQLite.cc) it looks like each row is inserted in its own transaction, but I could be mistaken. According to the [FAQ](https://www.sqlite.org/faq.html#q19), sqlite can handle tens of thousands of INSERTs per second, but only a few dozen transactions per second. In fact, when I try to manually insert the `conn.log` into a fresh sqlite file with `sqlitebrowser`, it happens basically instantaneously.

 So it seems plausible to me that this could be the reason, but someone who is more familiar with the code should confirm. If this is true, I suggest the INSERTs to be done in one single transaction.

For the record, I'm using `zeek version 3.2.3` built for Debian 11.

**Comment by JustinAzoff** [comment]
This is likely fsync overhead.  Are you on HDDs and not SSDs?  If you run zeek while in a tmpfs directory (check `mount | grep tmpfs`) does it run much faster?

The inserts can't really all be ran in a single transaction, but it should probably be batched.  There was some work being done to rework the zeek logging api to be more batch friendly, but I don't recall the status of that.

In the meantime there might be a few options.  If you are processing a ton of pcaps, you can probably do all the processing under tmpfs, and then just move the resulting databases to persistent storage when complete.  If you don't have enough free memory to use tmpfs, you could disable fsync in sqlite.  As long as you aren't worried about a sudden power failure during processing, there isn't much risk involved.  Something like WAL mode could also be an improvement.

**Comment by AdrianVollmer** [comment]
I was using an SSD.

You were right that when using a tmpfs it only took 20 seconds. Thanks for the hint!

**Comment by 0xxon** [comment]
I think we, from the beginning, were aware that the way that we do things for SQLite at the moment is not ideal from a speed perspective. From what I recall, we decide to do one transaction per insert to reduce the possibility of data loss during, e.g., crashes. 

I think SQLite was also not intended to be a backend that is used for high-volume logs. I can see an argument to change this to some kind of batching. We don't even really have to wait for this to be implemented on the Zeek side - this could be implemented completely on the reader with it using the heartbeats to end transactions.

Since this is kind of an extension request - we should perhaps convert this into a discussion as a possible extension to Zeek :)

**Comment by 0xxon** [comment]
This speed problem should be solved with the merge of #4063. The PR allows changing some SQLite configuration options, which can significantly enhance speed. 

Feel free to re-open the discussion if it doesn't solve the problem.
